### PR TITLE
Declaration emitter cleanup

### PR DESCRIPTION
**Pull request by mhegazy**
This is the first set of changes to get declaration bundling working. the first step is to split the emitter logic into two phases; the first to collect declarations that we need to emit (preprocess) and the second to write them (write). With these changes there should not be any functional changes, just error messages and reporting differences. I will be sending out changes to support --out with --module for declaration and then another set for bundling.

Note for reviewers, declarationEmitter.ts has changed significantly, so i would look at the file as whole instead of diffs. changes to other files are fairly limited though, so diffs should be sufficient.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -1657,7 +1513,7 @@ namespace ts {
                     // Unnamed function expressions, arrow functions, and unnamed class expressions have reserved names that
                     // we don't want to display
                     if (!isReservedMemberName(symbol.name)) {
-                        buildSymbolDisplay(symbol, writer, enclosingDeclaration, SymbolFlags.Type);
+                        buildSymbolDisplay(symbol, writer, enclosingDeclaration, SymbolFlags.Type, undefined, undefined, /*isTypeSymbol*/true);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Use named parameters here.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -1907,7 +1763,7 @@ namespace ts {
             }
 
             function buildTypeParameterDisplay(tp: TypeParameter, writer: SymbolWriter, enclosingDeclaration?: Node, flags?: TypeFormatFlags, symbolStack?: Symbol[]) {
-                appendSymbolNameOnly(tp.symbol, writer);
+                appendSymbolNameOnly(tp.symbol, writer, false);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Use named parameter

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -1922,7 +1778,7 @@ namespace ts {
                 if (isRestParameter(parameterNode)) {
                     writePunctuation(writer, SyntaxKind.DotDotDotToken);
                 }
-                appendSymbolNameOnly(p, writer);
+                appendSymbolNameOnly(p, writer, false);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Use named parameter

File: tests/baselines/reference/privacyTypeParameterOfFunctionDeclFile.errors.txt [hunk 4]
```diff
@@ -46,19 +46,19 @@ tests/cases/compiler/privacyTypeParameterOfFunctionDeclFile.ts(293,78): error TS
     export interface publicInterfaceWithPrivateTypeParameters {
         new <T extends privateClass>(): privateClass;  // Error
                        ~~~~~~~~~~~~
-!!! error TS4006: Type parameter 'T' of constructor signature from exported interface has or is using private name 'privateClass'.
+!!! error TS4002: Constraint of type parameter 'T' references inaccessible name 'privateClass'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/privacyTypeParameterOfFunctionDeclFile.errors.txt]: Much better

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -2325,4 +2324,13 @@ namespace ts {
             }
         }
     }
+
+    export function getTypeAnnotationFromAccessor(getAccessor: AccessorDeclaration, setAccessor: AccessorDeclaration): TypeNode {
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Is this used anywhere?

Comment (mhegazy) [inline src/compiler/utilities.ts]: declarationEmitter.ts

Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Shoot, GitHub can't show it.

**Comment by DanielRosenwasser** [comment]
In `declarationEmitter.ts`:

* Replace all "handel" with "handle"
* Replace all "infered" with "inferred"
* See where else we use lambdas that do nothing if you find any, let's see if we can move `emptyHandler` to `utilities.ts`.

**Comment by mhegazy** [comment]
closing for now. will refresh and resubmit.

**Pull request closed without merging**